src/s3.zig

```diff
@@ -0,0 +1,2110 @@
+        bun.http.http_thread.schedule(batch);
+    }
+
+    pub fn s3ReadableStream(this: *@This(), path: []const u8, offset: usize, size: ?usize, proxy_url: ?[]const u8, globalThis: *JSC.JSGlobalObject) JSC.JSValue {
+        var reader = JSC.WebCore.ByteStream.Source.new(.{
+            .context = undefined,
+            .globalThis = globalThis,
+        });
+
+        reader.context.setup();
+        const readable_value = reader.toReadableStream(globalThis);
+
+        this.s3StreamDownload(path, offset, size, proxy_url, @ptrCast(&S3DownloadStreamWrapper.callback), S3DownloadStreamWrapper.new(.{
+            .readable_stream_ref = JSC.WebCore.ReadableStream.Strong.init(.{
+                .ptr = .{ .Bytes = &reader.context },
+                .value = readable_value,
+            }, globalThis),
+        }));
+        return readable_value;
+    }
+
+    const S3DownloadStreamWrapper = struct {
+        readable_stream_ref: JSC.WebCore.ReadableStream.Strong,
+        pub usingnamespace bun.New(@This());
+
+        pub fn callback(chunk: bun.MutableString, has_more: bool, request_err: ?S3Error, this: *@This()) void {
+            defer if (!has_more) this.deinit();
+            if (this.readable_stream_ref.get()) |readable| {
+                if (readable.ptr == .Bytes) {
+                    const globalThis = this.readable_stream_ref.globalThis().?;
+                    const event_loop = globalThis.bunVM().eventLoop();
+                    event_loop.enter();
+                    defer event_loop.exit();
+                    if (request_err) |err| {
+                        readable.ptr.Bytes.onData(
+                            .{
+                                .err = .{ .JSValue = err.toJS(globalThis) },
+                            },
+                            bun.default_allocator,
+                        );
+                    } else if (has_more) {
+                        readable.ptr.Bytes.onData(
+                            .{
+                                .temporary = bun.ByteList.initConst(chunk.list.items),
+                            },
+                            bun.default_allocator,
+                        );
+                    } else {
+                        if (chunk.list.items.len == 0) {
+                            readable.ptr.Bytes.onData(
+                                .{
+                                    .done = {},
+                                },
+                                bun.default_allocator,
+                            );
+                        } else {
+                            readable.ptr.Bytes.onData(
+                                .{
+                                    .temporary_and_done = bun.ByteList.initConst(chunk.list.items),
+                                },
+                                bun.default_allocator,
+                            );
+                        }
+                    }
+                } else {
```

Comment:
can we instead make the scope above return, and then only put this code in one place instead of two?